SIS2 sea ice thickness and partial area relaxation

DmitryDukhovskoy opened the pull request:
Suggested changes include subroutines for implementing Newtonian relaxation of sea ice thickness and partial area (concentration) fields, which will relax the SIS2 predicted values toward the target values over a specified region with a prescribed relaxation rate.  The relaxation algorithm can serve as an alternative to the sea ice open boundary conditions in regional applications or when sea ice fields need to be nudged or corrected toward reference values within the domain. The main subroutines for initializing, processing, interpolating the target fields, and applying the relaxation are contained in the module SIS_sponge.F90. Necessary modifications have been made to several other SIS2 codes. To run the SIS2 code with relaxation, the user must turn SIS_SPONGE flag "on" (True), specify the relaxation parameters in SIS_input and provide an input file containing a 2D sea ice relaxation rate field, a 2D field of sea ice thicknesses (in meters), and the partial area. The relaxation code will then distribute ice thicknesses and partial areas by category.

DmitryDukhovskoy committed:
SIS2 ice relaxation code

DmitryDukhovskoy committed:
SIS2 ice thickness and partial area relaxataion codes

theresa-cordero commented on src/SIS_sponge.F90:
```diff
@@ -0,0 +1,989 @@
+  max_rlxrate =  maxval(Irelax*US%T_to_s)
+  call max_across_PEs(max_rlxrate)
+  call get_date(Time, year, mon, day, hr, minute, second, itick)
+  write(mesg,'("SIS Time:",i6,2("/",i2.2),1x,3(":",i2.2),"; max(Irelax)=",D13.4," s-1")') &
```
I like that you have added this sanity check, but I think it would be better to include logic to only write these extra messages if `DEBUG= True`.

DmitryDukhovskoy replied:
I'm not sure about this. I liked to have this option separate from the overall debug to be able to track the relaxation without printing out anything else. Having this information was very helpful to check the values with the target fields. Using "DEBUG" would complicate things for the users. Right now, it is very easy to add/not this information by simply specifying the test point in the SIS_input. Adding debug will require the user to remember to turn this flag on in addition to adding the test point, and if the test point is not needed, then the user will have to remember to turn the debug flag off. Besides, having debug on, will add lots of unneeded information if the only purpose is to check ice relaxation.

marshallward replied:
These messages appear to always be printed, is that right?  I see a lot of `SIS_mesg()` calls in the function.

We generally try to minimize diagnostic output like this during runtime.  If every module were to report its status, then warnings and other serious messages could be easily missed.

One solution is to add a verbosity level to `SIS_mesg(message, verb)`.  The default value is 2, so anything higher will be removed.  A higher value, say 5, is sometimes used to print diagnostic info.  `DEBUG` corresponds to 9, which prints everything.

Hopefully this avoids any questions about complex juggling of multiple `DEBUG*` flags.

You might also want to wrap the `max_across_PEs` around a verbosity flag.  It's always a good idea to avoid collectives if you can manage it.

(Unrelated point, but `.lt.` and `.gt.` should also be replace with `<` and `>` when you get a chance.)

DmitryDukhovskoy replied:
Thanks Marshall, I agree.  With the test point information, would this be ok if I leave it as is? It is easy to turn on/off this diagnostics in SIS_input by specifying or not the test point indices. Otherwise it would require the user to specify both the test point and verbosity level. Or how would you do this?

marshallward replied:
Using `ISPONGE_[IJ]TEST` to control output seems ok to me, although I don't know if there's any precedent for this sort of output.

The closest I can think of is truncation reporting in (say) `MOM_vert_friction.F90` or `MOM_PointAccel.F90` which logs invalid states to a file, rather than monitoring a designated point.

I'd say leave it for now but let's see what @Hallberg-NOAA thinks when he returns next week.

DmitryDukhovskoy replied:
Sounds good, thanks! In the latest commit, I think I addressed all comments including "max_across_PEs".

theresa-cordero commented on src/SIS_sponge.F90:
```diff
@@ -0,0 +1,989 @@
+                 "The exact location and properties of those sponges are "//&
+                 "specified from MOM_initialization.F90.", default=.false.)
+
+  if (use_isponge) then
```
I think these messages are redundant because the SIS parameter file will have SIS_SPONGE=True or False

DmitryDukhovskoy replied:
Sure, this can be deleted. This is a leftover from my development version of the code. I wanted to have all information in the log file.

theresa-cordero commented:
Thanks for this great addition to SIS2! 

I have taken an initial look at the code and made a couple minor suggests. I also asked @Hallberg-NOAA and @marshallward to take a closer look at the code, especially the IO.

DmitryDukhovskoy commented:
Theresa, thank you for looking into the code. I'll wait for comments from others to make all final changes.

kshedstrom commented:
How about a DEBUG_SPONGE option?

DmitryDukhovskoy commented:
> How about a DEBUG_SPONGE option?

Sure, this is an option. Still the user will need to specify the test point where he/she wants the diagnostics to be printed out, right?  Does this mean that if the test points are provided but DEBUG_SPONGE=False, then no diagnostics are printed out (the test points ignored), similarly if the DEBUG_SPONGE=True but no test point is specified - no diagnostics. Right?

kshedstrom commented:
If you write it that way, sure.

DmitryDukhovskoy committed:
Added verbosity to SIS_mesg, changed .gt. .lt. to > and <

Hallberg-NOAA commented on src/SIS_slow_thermo.F90:
```diff
@@ -59,6 +59,8 @@ module SIS_slow_thermo
 use SIS2_ice_thm,      only : enth_from_TS, Temp_from_En_S, enthalpy_liquid, calculate_T_freeze
+use SIS_sponge,        only : initialize_icerelax_file, set_up_isponge_field, apply_isponge, isponge_CS
+use SIS_sponge,        only : global_to_local_ij
 
```
`global_to_local_ij()` and `set_up_isponge_feild` are not used in SIS_slow_thermo.F90, and hence should be remove from these `use ..., only` statements.

DmitryDukhovskoy replied:
Thank you, missed that. Done.

Hallberg-NOAA commented on src/SIS_sponge.F90:
```diff
@@ -0,0 +1,991 @@
+  integer :: mon, hr, minute, itick
+  integer :: start_of_day, num_days
+  integer :: verbosity   !< MOM verbosity level
+  real :: max_rlxrate, rho_ice
```
Please add comments describing the real variables `max_rlxrate` and `rho_ice` along with their units.

DmitryDukhovskoy replied:
The description has been added

Hallberg-NOAA commented on src/SIS_sponge.F90:
```diff
@@ -0,0 +1,991 @@
+use SIS_diag_mediator, only : post_SIS_data, post_data=>post_SIS_data
+use SIS_diag_mediator, only : register_diag_field=>register_SIS_diag_field
+use SIS_hor_grid,      only : SIS_hor_grid_type
+use SIS_sum_output,    only : SIS_sum_out_CS, write_ice_statistics! , SIS_sum_output_init
```
I don't believe that `write_ice_statistics()` or `SIS_sum_out_CS` are used in the SIS_sponge.F90 module.

DmitryDukhovskoy replied:
Corrected

Hallberg-NOAA commented on src/SIS_sponge.F90:
```diff
@@ -0,0 +1,991 @@
+  call MOM_read_data(filename, rlxrate_var, Irelax(:,:), G%Domain, scale=US%s_to_T)
+  max_rlxrate =  maxval(Irelax*US%T_to_s)
+  call get_date(Time, year, mon, day, hr, minute, second, itick)
+  call get_time(Time, start_of_day, num_days)
```
I do not believe that `start_of_day` or `num_days` are used.

DmitryDukhovskoy replied:
This is correct, I also found this while working on other changes in the code. get_time is not needed.

Hallberg-NOAA commented on src/SIS_sponge.F90:
```diff
@@ -0,0 +1,991 @@
+    call SIS_error(FATAL, " initialize_icerelax_file: Unable to open "//trim(filename))
+
+  call MOM_read_data(filename, rlxrate_var, Irelax(:,:), G%Domain, scale=US%s_to_T)
+  max_rlxrate =  maxval(Irelax*US%T_to_s)
```
Please move lines 176 and 177 setting `max_rlxrate` and the time variables into the `if (verb_msg > verbosity)` block, as they are not used otherwise.

DmitryDukhovskoy replied:
Good observation, max_rlxrate and get_date(Time, ...) moved into the block.  Also get_time(Time, start_of_day, num_days) is not needed and has been deleted.

Hallberg-NOAA commented on src/SIS_sponge.F90:
```diff
@@ -0,0 +1,991 @@
+  CS%use_isponge = use_isponge
+  if (present(itest) .and. present(jtest)) then
+    write(mesg,'(A," itest/jtest =",2(i5,1x))') trim(mdl), itest, jtest
+    write(*,'(A)') trim(mesg)
```
Using `write(*,...)` from all PEs can generate a very large volume of redundant information.  Instead, use `SIS_mesg(msg)` here to only write this message from the root PE.

marshallward replied:
@Hallberg-NOAA I believe this reports on an individual grid point so would only be printed from a single PE.

(But it probably can't hurt to use `SIS_mesg`.)

DmitryDukhovskoy replied:
I totally agree — I also cringed when I used write(*,...) for debugging. However, I couldn’t come up with a better solution at the time. Marshall is right: the message is reported only for a specific grid point on a single PE. I did try using SIS_mesg, but it didn’t work. I’ll give it another shot. Happy to implement a better approach if you have any suggestions.

Hallberg-NOAA replied:
Try using `call SIS_mesg(mesg, all_print=.true.)`  Without the `all_print` argument, only the root PE writes this message.

DmitryDukhovskoy replied:
Hi @Hallberg-NOAA, it seemed like a right approach, unfortunately it didn't work.  I implemented this change in the code and all messages disappeared. Then I also tried doing this:   

  CS%use_isponge = use_isponge
  if (present(itest) .and. present(jtest)) then
    write(mesg,'(A," itest/jtest =",2(i5,1x))') trim(mdl), itest, jtest  
    write(*,'(A)') trim(mesg)             ! <--- Old version
    call SIS_mesg(trim(mesg), all_print=.true.)  ! <--- Suggested change

In the log file, I'm getting the message only from my version and none from SIS_mesg. 
initialize_isponge itest/jtest =   13     6

I also tried this to see if all_print works:
  write(mesg,'(A,": SIS_SPONGE IS ON")') trim(mdl)
  call SIS_mesg(mesg, all_print=.true.)

Checking the MOM_mesg subroutine, I would expect to see this message reported from all PEs, however, I'm getting only one message:
NOTE from PE     0: initialize_isponge: SIS_SPONGE IS ON

I'm not sure why all_print option does not work, it should.

Hallberg-NOAA commented on src/SIS_sponge.F90:
```diff
@@ -0,0 +1,991 @@
+  integer :: CatIce                  !< number of sea ice categories.
+  integer :: icat0                   !< cat where to assign original hice and cice to begin distr.
+
+  real, allocatable, dimension(:,:) :: cice2d, hice2d
```
Please add comments describing each of the real variables in this routine, along with their units.

DmitryDukhovskoy replied:
Comments have been added.

Hallberg-NOAA commented on src/SIS_sponge.F90:
```diff
@@ -0,0 +1,991 @@
+!< Check if total ice thkn*conc and conc are conserved (i.e. equal original hice, cice)
+subroutine check_hcice(CatIce, hcat, ccat, hice, cice, err_hice, err_cice, &
+                       hice_tot, cice_tot, iG, jG, str, verb)
+  integer, intent(in) :: CatIce
```
SIS2 mostly follows the MOM6 style guide, so all subroutine arguments are required to be described in doxygen syntax comments, including units for real variables.  This same comment applies to several of the following routines.

DmitryDukhovskoy replied:
Corrected

Hallberg-NOAA commented on src/SIS_sponge.F90:
```diff
@@ -0,0 +1,991 @@
+#include <SIS2_memory.h>
+
+public initialize_icerelax_file, apply_isponge, set_up_isponge_field, SIS_sponge_end
+public global_to_local_ij
```
I do not believe that `global_to_local_ij` is used outside of this module. If so it need not be declared as public.

DmitryDukhovskoy replied:
Corrected.

marshallward commented on src/SIS_sponge.F90:
```diff
@@ -0,0 +1,991 @@
+    enddo
+
+    call partial_area_total(CatIce, ccat(1:CatIce), cice_tot)
+    call ice_thkn_total(CatIce, ccat(1:CatIce), hcat, hice_tot)
```
`ccat` is allocated as size `CatIce` in L704, so using `ccat(1:CatIce)` seems redundant.  This could also cause an implicit copy inside of the function.  Can `ccat` be used here?

DmitryDukhovskoy replied:
Marshall, you are probably right. I noticed this is the way it is done in several SIS2 codes, e.g. 
https://github.com/DmitryDukhovskoy/SIS2_irelax/blob/2c49005d1d7132776fff9b501cabe6ae85ef5f57/src/ice_ridge.F90#L309
The issue is that some arrays are allocated as (0:CatIce) while others use (1:CatIce). It's very easy to make a mistake and assign the wrong arrays. I encountered this myself when I used ccat = IST%part_size(i,j,:), where part_size was declared as (0:CatIce). It took me a while to figure out why the relaxation wasn't working—there was no compilation error.

Although it's a bit redundant, explicitly specifying dimensions makes the code safer. That’s why several SIS codes follow this approach. I'm happy to change the code back (as I had it before and as you're suggesting) if that is the preferred style.

marshallward replied:
You have a good point, the start index of ice category changes often enough that it's worth calling it out.

I am mainly concerned that it could trigger a redundant copy, e.g.
```
tmp(:) = ccat(1:CatIce)
call partial_area_total(CatIce, tmp, cice_tot)
```
It's hard to know what a compiler will do in this case.  If the input is `intent(in)` and it knows that `CatIce` won't change, then the code output is probably identical.  But I would feel more comfortable if it were not given the option to copy.

Another way to emphasize the one-indexing is to make it explicit in the allocation?
```
allocate(ccat(1:CatIce))
```
This information could also be provided in the variable description.

DmitryDukhovskoy replied:
I got it, you are right. I'll make changes in the code. Thanks!

marshallward replied:
I looked into this a little more, and I was being overly cautious.  `ccat(1:CatIce)` is considered "simply contiguous", regardless of the size of `ccat`, which ensures that it will not produce a temporary copy.

Sorry about making this an issue.  I guess we're all still learning every day :flushed: But thank you for making the change.

DmitryDukhovskoy replied:
Hi @marshallward, no worries. I suppose we want to be conservative with code development. I wasn’t thinking about temporary copies when I originally wrote this, and your suggestion seems reasonable.

marshallward commented:
Hi @DmitryDukhovskoy
Several of the functions (`find_icat`, `partial_area_total`, `ice_thnk_total`, `check_hice`) were written with a size argument `CatIce`. This could be removed by passing the arguments as deferred size, e.g.
```
real, intent(in) :: cice_cat(:)
```
and using `size()` to set the loop bounds.

Not urgent, but if you are making other revisions then this could be included.

DmitryDukhovskoy commented:
> Hi @DmitryDukhovskoy Several of the functions (`find_icat`, `partial_area_total`, `ice_thnk_total`, `check_hice`) were written with a size argument `CatIce`. This could be removed by passing the arguments as deferred size, e.g.
> 
> ```
> real, intent(in) :: cice_cat(:)
> ```
> 
> and using `size()` to set the loop bounds.
> 
> Not urgent, but if you are making other revisions then this could be included.

Hi @marshallward , thank for the suggestion. It is done.

DmitryDukhovskoy committed:
Added missing descriptions, deleted unneeded routines, improved clarity of subroutines operating on arrays following dev. group suggestions

DmitryDukhovskoy commented:
Hi @Hallberg-NOAA and @marshallward, I think I've addressed all your comments in the recent commit, except for Bob's comment regarding write(*,'(A)') trim(mesg) for the test point. I could not come up with a better solution. I'm ok to remove the "test point" option from the code to resolve this problem. The "test point" was very helpful during code development but perhaps it is not so useful for the user. Let me know. I compiled the updated code and ran several test simulations to verify the code performance. See attached figure showing ice thickness fields from the simulations with varying relaxation rates and the target fields (at the top).  Thank you for your suggestions and time. 
<img width="1958" alt="Screenshot 2025-05-02 at 3 04 40 PM" src="https://github.com/user-attachments/assets/249cfef0-87bf-4272-8b1f-7909a354cc9f" />

theresa-cordero commented on src/SIS_sponge.F90:
```diff
@@ -0,0 +1,973 @@
+                 "ISPONGE_STATE_FILE.", default="ithkn")
+  call get_param(param_file, mdl, "ISPONGE_IAREA_VAR", iarea_var, &
+                 "The name of the ice partial area/concentration variable in "//&
+                 "ISPONGE_STATE_FILE.", default="ithkn")
```
Should the default for `ISPONGE_IAREA_VAR` be different than the default for `ISPONGE_ITHCK_VAR`?

DmitryDukhovskoy replied:
It should, good catch. Thank you. I modified the code, made sure it compiled, committed and pushed the changes.

DmitryDukhovskoy committed:
Corrected default name for partial area variable in the target file

DmitryDukhovskoy committed:
Fixed a bug in the implementation of the SIS_SPONGE=False option

marshallward committed:
Merge branch 'dev/gfdl' into gaea_irlx

marshallward commented:
@DmitryDukhovskoy We've had a fair bit of debate on our end about the `[ij]test` feature, and how to integrate into the codebase.  It looks like it has been an invaluable tool during development of the SIS sponge, and a lot of careful work has been put into its design.  But there are problems we need to consider.

* `write()` is a problematic command for a few reasons:

  * It lacks parallel control.  There is the burden of working out which rank is the I/O source.

  * Generally we cede I/O to FMS, rather than doing directly (although exceptions can be found).  It is possible to get a `stdout` unit from FMS, but this is typically only done at the driver level.

  * `write()` on GPU would raise an error, if we ever get to that point.

* Several `if (CS%itest > 0)` tests are within do-loops, which can impair vectorization and the overall performance of model loops.

While I imagine that we could work through these problems, it could take time to reach a point which satisfies everyone.

I know it's probably not what you would like to hear, but how receptive would you be to removing the `[ij]test` feature?

DmitryDukhovskoy commented:
> @DmitryDukhovskoy We've had a fair bit of debate on our end about the `[ij]test` feature, and how to integrate into the codebase. It looks like it has been an invaluable tool during development of the SIS sponge, and a lot of careful work has been put into its design. But there are problems we need to consider.
> 
> * `write()` is a problematic command for a few reasons:
>   
>   * It lacks parallel control.  There is the burden of working out which rank is the I/O source.
>   * Generally we cede I/O to FMS, rather than doing directly (although exceptions can be found).  It is possible to get a `stdout` unit from FMS, but this is typically only done at the driver level.
>   * `write()` on GPU would raise an error, if we ever get to that point.
> * Several `if (CS%itest > 0)` tests are within do-loops, which can impair vectorization and the overall performance of model loops.
> 
> While I imagine that we could work through these problems, it could take time to reach a point which satisfies everyone.
> 
> I know it's probably not what you would like to hear, but how receptive would you be to removing the `[ij]test` feature?

@marshallward, thanks for taking the time to work on this problem. These are all important concerns. I like this feature, but it’s probably not that critical for the user unless ice nudging needs debugging. I'm ok to remove the ij-test feature. I'll make the necessary changes in the code.

DmitryDukhovskoy committed:
Removed ij-test feature for ice relaxation debugging

DmitryDukhovskoy commented:
Hi, The updated version of the code with the ij-test debugging feature removed has been pushed. I’ve tested this version, and everything appears to be working correctly; I haven’t encountered any issues so far. I do regret having to remove the ij-test feature, as I relied on it heavily while evaluating recent code performance and resolving issues. However, I understand the complications it introduced and why it needed to be removed from the shared version.
For now, I’ll maintain a separate version with the ij-test functionality for my own use. 
Thanks!

marshallward reviewed (dismissed):
@DmitryDukhovskoy Thanks for your understanding.

I can see the value in a per-element diagnostic, it feels like a gap in our existing tools.  But it may be easier to evaluate as a separate PR.

I think all of the issues have been addressed so I will run this through the regression suite and, assuming it passes, will merge it.

marshallward committed:
SIS_sponge: IO width fix

Add a width to the integer output in SIS_Sponge.F90

marshallward commented:
Gaea regression: https://gitlab.gfdl.noaa.gov/ogrp/mom6ci/SIS2/-/pipelines/27659 :heavy_check_mark: :yellow_circle:
